### PR TITLE
Add completion status

### PR DESCRIPTION
**Pull request by cwiklik**

### CODE REVIEW

File: pkg/apis/controller/v1beta1/appwrapper.go [hunk 1]
```diff
@@ -266,11 +267,24 @@ type AppWrapperStatus struct {
...
+
+	ItemCompletionStatus GenericItemCompletionStatus `json:"itemCompletionStatus,omitempty"`
+}
+
+type GenericItemCompletionStatus struct {
+	GenericItems [] GenericItem `json:"genericItems,omitempty"`
+}
+
+type GenericItem struct {
+	schema.GroupVersionKind `json:"gvk"`
```
Comment (dmatch01) [inline pkg/apis/controller/v1beta1/appwrapper.go:279]: Can you reach out to Alexey to see if this field works for him.  Consumers of this field will need to match gvk. Would like to get feed back before merge.

File: config/crd/bases/mcad.ibm.com_appwrappers.yaml [hunk 2]
```diff
@@ -889,6 +889,38 @@ spec:
...
+                properties:
+                  genericItems:
+                    items:
+                      properties:
+                        condition:
+                          type: string
+                        gvk:
+                          description: Group Version Kind 
+                          properties:
+                            Group:
```
Comment (dmatch01) [inline config/crd/bases/mcad.ibm.com_appwrappers.yaml:902]: readability suggestion: order this based on field name above `gvk`.

File: deployment/mcad-controller/crds/mcad.ibm.com_appwrappers.yaml [hunk 3]
```diff
@@ -889,6 +889,38 @@ spec:
...
+                          - Version
+                          type: object  
+                        name:
+                          type: string
+                        namespace:
+                          type: string
+                      required:
+                      - condition
+                      - name
+                      - namespace
```
Comment (dmatch01) [inline deployment/mcad-controller/crds/mcad.ibm.com_appwrappers.yaml:920]: Some objects may not be namespaced so this field may not exist.  An example is a `namespace` and `clusterrolebinding`.  While those objects are not specific to a completion status there are objects that a may fit the bill and we would want to support that model.

File: deployment/mcad-controller/crds/mcad.ibm.com_appwrappers.yaml [hunk 4]
```diff
@@ -889,6 +889,38 @@ spec:
...
                 format: int32
                 type: integer
+              itemCompletionStatus:
+                properties:
+                  genericItems:
+                    items:
+                      properties:
+                        condition:
+                          type: string
+                        gvk:
```
Comment (dmatch01) [inline deployment/mcad-controller/crds/mcad.ibm.com_appwrappers.yaml:899]: This field should probably be part of the required.

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 5]
```diff
@@ -695,21 +697,45 @@ func (qjm *XController) getAppWrapperCompletionStatus(caw *arbv1.AppWrapper) arb
 			}
 			unstruct.Object = blob.(map[string]interface{}) //set object to the content of the blob after Unmarshalling
 			name := ""
+			namespace := ""
 			if md, ok := unstruct.Object["metadata"]; ok {
 				metadata := md.(map[string]interface{})
 				if objectName, ok := metadata["name"]; ok {
 					name = objectName.(string)
 				}
+				if objectName, ok := metadata["namespace"]; ok {
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go:706]: Please add a test case for non-namedspaced test case to ensure a nil here would not panic.  Thanks!

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 6]
```diff
@@ -695,21 +697,45 @@ func (qjm *XController) getAppWrapperCompletionStatus(caw *arbv1.AppWrapper) arb
...
+					namespace = objectName.(string)
+				}
 			}
+			_, gvk, err := unstructured.UnstructuredJSONScheme.Decode(objectName.Raw, nil, nil)
+			if err != nil {
+				klog.Errorf("[getAppWrapperCompletionStatus] Error unmarshalling, err=%#v", err)
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go:712]: Suggest a value of `Unknown` if error?

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 7]
```diff
@@ -695,21 +697,45 @@ func (qjm *XController) getAppWrapperCompletionStatus(caw *arbv1.AppWrapper) arb
...
 			klog.Infof("[getAppWrapperCompletionStatus] Checking items completed for appwrapper: %v in namespace: %v", caw.Name, caw.Namespace)
 
-			status := qjm.genericresources.IsItemCompleted(&genericItem, caw.Namespace, caw.Name, name)
+			status, condition := qjm.genericresources.IsItemCompleted(&genericItem, caw.Namespace, caw.Name, name)
 			if !status {
 				//early termination because a required item is not completed
-				return caw.Status.State
+				return caw.Status.State, completionStatus
+			} else {
+				_ = condition
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go:725]: Q: what does this statement do?

**Review by dmatch01 — APPROVED** [review]
Unit test would be great, e2e test would be min. at the least.

File: pkg/controller/queuejob/queuejob_controller_ex.go (hunk 7 above)
Comment (cwiklik) [reply pkg/controller/queuejob/queuejob_controller_ex.go:725]: Ah, no longer need it. It just a way to silence go "unused var" error.

File: pkg/apis/controller/v1beta1/appwrapper.go (hunk 1 above)
Comment (cwiklik) [reply pkg/apis/controller/v1beta1/appwrapper.go:279]: I reached out to Alexey but he ran out of time to address it today. Will be back on Sunday. Should we wait with merge till Monday?

Comment (cwiklik) [reply pkg/apis/controller/v1beta1/appwrapper.go:279]: Alexey ran out of time today to address this. He will be back on Sunday. He had some comments

File: pkg/controller/queuejob/queuejob_controller_ex.go (hunk 5 above)
Comment (cwiklik) [reply pkg/controller/queuejob/queuejob_controller_ex.go:706]: if objectNamespace, ok := metadata["namespace"]; ok {
					if objectNamespace != nil {
						namespace = objectNamespace.(string)
					}
				}

**Comment by cwiklik** [comment]
Not viable
